### PR TITLE
Add backend e2e tests

### PR DESCRIPTION
**Pull request by adinauer**
#skip-changelog

## :scroll: Description

Add some end to end tests for backend.

## :bulb: Motivation and Context

To catch breaking changes early.

## :green_heart: How did you test it?
;-)

## :pencil: Checklist

- [ ] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -15,6 +15,10 @@ clean:
 compile:
 	./gradlew build
 
+# build without tests
+compileWithoutTests:
```
Comment (romtsn) [inline Makefile:19]: that's probably not needed anymore?

File: .github/workflows/system-tests-backend.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
+name: system-tests-backend
+
+on:
+  push:
+    branches:
+      - main
+      - release/**
```
Comment (romtsn) [inline .github/workflows/system-tests-backend.yml:7]:
```suggestion
```
as discussed privately, let's only run it for `main` and PRs

File: .github/workflows/system-tests-backend.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.10.5'
+
+      - name: Set up Java
+        uses: actions/setup-java@v4
+        with:
+          distribution: 'temurin'
+          java-version: ${{ matrix.java }}
```
Comment (romtsn) [inline .github/workflows/system-tests-backend.yml:43]:
```suggestion
          java-version: '17'
```
could get rid of the java version matrix in theory

File: .github/workflows/system-tests-backend.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+      - name: Exclude android modules from build
+        run: |
+          sed -i -e '/.*"sentry-android-ndk",/d' -e '/.*"sentry-android",/d' -e '/.*"sentry-compose",/d' -e '/.*"sentry-android-core",/d' -e '/.*"sentry-android-fragment",/d' -e '/.*"sentry-android-navigation",/d' -e '/.*"sentry-android-okhttp",/d' -e '/.*"sentry-android-sqlite",/d' -e '/.*"sentry-android-timber",/d' -e '/.*"sentry-android-integration-tests:sentry-uitest-android-benchmark",/d' -e '/.*"sentry-android-integration-tests:sentry-uitest-android",/d' -e '/.*"sentry-android-integration-tests:test-app-sentry",/d' -e '/.*"sentry-samples:sentry-samples-android",/d' settings.gradle.kts
+
+      - name: Exclude android modules from ignore list
+        run: |
+          sed -i -e '/.*"sentry-uitest-android",/d' -e '/.*"sentry-uitest-android-benchmark",/d' -e '/.*"test-app-sentry",/d' -e '/.*"sentry-samples-android",/d' build.gradle.kts
+
+      - name: Start server and run integration test for sentry-cli commands
```
Comment (romtsn) [inline .github/workflows/system-tests-backend.yml:56]:
```suggestion
      - name: Build server jar
```

**Review by romtsn — APPROVED** [review]
a few minor things and I didn't check the test code, only configs and gradle files, but LGTM otherwise 🚢

**Comment by adinauer** [comment]
Actuator has only been added to one of the sample projects and should be there for all so the ...wait for spring ... script doesn't wait the full duration but stops when the server is ready.

**Comment by adinauer** [comment]
JUNIT test reports seem to override each other

File: sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.apollographql.apollo3.ApolloClient
+import com.apollographql.apollo3.api.ApolloResponse
+import com.apollographql.apollo3.api.Mutation
+import com.apollographql.apollo3.api.Query
+import io.sentry.samples.graphql.GreetingQuery
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+
+class GraphqlTestClient(val backendUrl: String) {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt:11]: (l) `val` could be ommitted, backendUrl is only used to iniitialize `apollo`

File: sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    companion object {
+        val logger = LoggerFactory.getLogger(GraphqlTestClient::class.java)
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt:14]: (l) companion object could be deleted, `logger` is unused

File: sentry-samples/sentry-samples-spring-boot-jakarta/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    @Test
+    fun `get todo works`() {
+        val envelopeCountBefore = testHelper.sentryClient.getEnvelopeCount()
+        println(envelopeCountBefore)
+
+        val restClient = testHelper.restClient
+        restClient.getTodo(1L)
+        assertEquals(HttpStatus.OK, restClient.lastKnownStatusCode)
+
+        Thread.sleep(1000)
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-jakarta/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt:28]: can be removed if `testHelper.ensureEnvelopCountIncreased()` is used

File: sentry-samples/sentry-samples-spring-boot-jakarta/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    @Test
+    fun `get todo webclient works`() {
+        val envelopeCountBefore = testHelper.sentryClient.getEnvelopeCount()
+        println(envelopeCountBefore)
+
+        val restClient = testHelper.restClient
+        restClient.getTodoWebclient(1L)
+        assertEquals(HttpStatus.OK, restClient.lastKnownStatusCode)
+
+        Thread.sleep(1000)
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-jakarta/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt:44]: can be removed if `testHelper.ensureEnvelopCountIncreased()` is used

File: sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/PersonSystemTest.kt [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    lateinit var testHelper: TestHelper
+
+    @Before
+    fun setup() {
+        testHelper = TestHelper("http://localhost:8080")
+    }
+
+    @Test
+    fun `get person fails`() {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/PersonSystemTest.kt:21]: See comments in `samples-spring-boot-jakarta` version of this test

File: sentry-samples/sentry-samples-spring-boot-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import com.apollographql.apollo3.api.Mutation
+import com.apollographql.apollo3.api.Query
+import io.sentry.samples.graphql.AddProjectMutation
+import io.sentry.samples.graphql.GreetingQuery
+import io.sentry.samples.graphql.ProjectQuery
+import io.sentry.samples.graphql.TasksAndAssigneesQuery
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+
+class GraphqlTestClient(val backendUrl: String) {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt:14]: (l) val could be ommitted, backendUrl is only used to iniitialize apollo

File: sentry-samples/sentry-samples-spring-boot-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+    companion object {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt:16]: (l) companion object could be deleted, logger is unused

File: sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+    lateinit var testHelper: TestHelper
+
+    @Before
+    fun setup() {
+        testHelper = TestHelper("http://localhost:8080")
+    }
+
+    @Test
+    fun `get todo webclient works`() {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt:20]: See comments in `samples-spring-boot-jakarta` version of this test

File: sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/util/TestHelper.kt [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        restClient = RestTestClient(backendUrl)
+        sentryClient = SentryMockServerClient("http://localhost:8000")
+        graphqlClient = GraphqlTestClient(backendUrl)
+    }
+
+    fun snapshotEnvelopeCount() {
+        envelopeCounts = sentryClient.getEnvelopeCount()
+    }
+
+    fun ensureEnvelopCountIncreased() {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/util/TestHelper.kt:28]: small typeo, should be `ensureEnvelopeCountIncreased`

File: sentry-samples/sentry-samples-spring-boot-webflux/src/test/kotlin/io/sentry/systemtest/PersonSystemTest.kt [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+    lateinit var testHelper: TestHelper
+
+    @Before
+    fun setup() {
+        testHelper = TestHelper("http://localhost:8080")
+    }
+
+    @Test
+    fun `get person fails`() {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux/src/test/kotlin/io/sentry/systemtest/PersonSystemTest.kt:21]: See comments in `samples-spring-boot-jakarta` version of this test

File: sentry-samples/sentry-samples-spring-boot-webflux/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+    lateinit var testHelper: TestHelper
+
+    @Before
+    fun setup() {
+        testHelper = TestHelper("http://localhost:8080")
+    }
+
+    @Test
+    fun `get todo webclient works`() {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt:20]: See comments in `samples-spring-boot-jakarta` version of this test

File: sentry-samples/sentry-samples-spring-boot-webflux/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.apollographql.apollo3.ApolloClient
+import com.apollographql.apollo3.api.ApolloResponse
+import com.apollographql.apollo3.api.Mutation
+import com.apollographql.apollo3.api.Query
+import io.sentry.samples.graphql.GreetingQuery
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+
+class GraphqlTestClient(val backendUrl: String) {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt:11]: See comments in `samples-spring-boot-jakarta` version of this class

File: sentry-samples/sentry-samples-spring-boot-webflux/src/test/kotlin/io/sentry/systemtest/util/TestHelper.kt [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+        restClient = RestTestClient(backendUrl)
+        sentryClient = SentryMockServerClient("http://localhost:8000")
+        graphqlClient = GraphqlTestClient(backendUrl)
+    }
+
+    fun snapshotEnvelopeCount() {
+        envelopeCounts = sentryClient.getEnvelopeCount()
+    }
+
+    fun ensureEnvelopCountIncreased() {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux/src/test/kotlin/io/sentry/systemtest/util/TestHelper.kt:28]: small typeo, should be: `ensureEnvelopeCountIncreased`

File: sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/PersonSystemTest.kt [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+    lateinit var testHelper: TestHelper
+
+    @Before
+    fun setup() {
+        testHelper = TestHelper("http://localhost:8080")
+    }
+
+    @Test
+    fun `get person fails`() {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/PersonSystemTest.kt:21]: See comments in `samples-spring-boot-jakarta` version of this test

File: sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+    lateinit var testHelper: TestHelper
+
+    @Before
+    fun setup() {
+        testHelper = TestHelper("http://localhost:8080")
+    }
+
+    @Test
+    fun `get todo works`() {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt:20]: See comments in `samples-spring-boot-jakarta` version of this test

File: sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+        Thread.sleep(1000)
+
+        val envelopeCountAfter = testHelper.sentryClient.getEnvelopeCount()
+        println(envelopeCountAfter)
+        assertTrue(envelopeCountAfter.envelopes!! > envelopeCountBefore.envelopes!!)
+    }
+
+    @Test
+    fun `get todo webclient works`() {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/TodoSystemTest.kt:36]: See comments in `samples-spring-boot-jakarta` version of this test

File: sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/util/TestHelper.kt [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+        restClient = RestTestClient(backendUrl)
+        sentryClient = SentryMockServerClient("http://localhost:8000")
+        graphqlClient = GraphqlTestClient(backendUrl)
+    }
+
+    fun snapshotEnvelopeCount() {
+        envelopeCounts = sentryClient.getEnvelopeCount()
+    }
+
+    fun ensureEnvelopCountIncreased() {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/util/TestHelper.kt:28]: small typeo, should be: `ensureEnvelopeCountIncreased`

File: sentry-samples/sentry-samples-spring-boot-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt [hunk 22]
```diff
@@ -0,0 +1,48 @@
+package io.sentry.systemtest.grahql
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt:1]: typeo in package name, should be `io.sentry.systemtest.graphql`

File: sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt [hunk 23]
```diff
@@ -0,0 +1,33 @@
+package io.sentry.systemtest.grahql
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux-jakarta/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt:1]: typeo in package name, should be `io.sentry.systemtest.graphql`

File: sentry-samples/sentry-samples-spring-boot-webflux/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt [hunk 24]
```diff
@@ -0,0 +1,33 @@
+package io.sentry.systemtest.grahql
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot-webflux/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt:1]: typeo in package name, should be `io.sentry.systemtest.graphql`

File: sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt [hunk 25]
```diff
@@ -0,0 +1,48 @@
+package io.sentry.systemtest.grahql
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt:1]: typeo in package name, should be `io.sentry.systemtest.graphql`

File: sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+import com.apollographql.apollo3.api.Mutation
+import com.apollographql.apollo3.api.Query
+import io.sentry.samples.graphql.AddProjectMutation
+import io.sentry.samples.graphql.GreetingQuery
+import io.sentry.samples.graphql.ProjectQuery
+import io.sentry.samples.graphql.TasksAndAssigneesQuery
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+
+class GraphqlTestClient(val backendUrl: String) {
```
Comment (lbloder) [inline sentry-samples/sentry-samples-spring-boot/src/test/kotlin/io/sentry/systemtest/grahql/GraphqlTestClient.kt:14]: See comments in `samples-spring-boot-jakarta` version of this class

**Review by lbloder — APPROVED** [review]
LGTM 👍

**Pull request merged**